lib/class-wp-annotation-utils.php

```diff
@@ -0,0 +1,818 @@
+	 * Otherwise, it closely resembles {@see wp_unspam_comment()}.
+	 *
+	 * @since [version]
+	 *
+	 * @param  WP_Comment|int $comment Comment (i.e., annotation) object or ID.
+	 *
+	 * @return bool                    True on success. False on failure.
+	 *
+	 * @see wp_unspam_comment()
+	 */
+	public static function unspam_comment( $comment ) {
+		if ( ! $comment ) {
+			return false;
+		}
+
+		$comment = get_comment( $comment );
+
+		if ( ! $comment ) {
+			return false;
+		}
+
+		/** This action is documented in wp-includes/comment.php */
+		do_action( 'unspam_comment', $comment->comment_ID, $comment );
+
+		$status = (string) get_comment_meta( $comment->comment_ID, '_wp_trash_meta_status', true );
+		$status = $status ? $status : '0';
+
+		if ( self::set_comment_status( $comment, $status ) ) {
+			delete_comment_meta( $comment->comment_ID, '_wp_trash_meta_status' );
+			delete_comment_meta( $comment->comment_ID, '_wp_trash_meta_time' );
+
+			/** This action is documented in wp-includes/comment.php */
+			do_action( 'unspammed_comment', $comment->comment_ID, $comment );
+
+			return true;
+		}
+
+		return false;
+	}
+
+	/**
+	 * Removes a comment from trash (supports custom restoration statuses).
+	 *
+	 * Unlike {@see wp_untrash_comment()}, this supports custom restoration statuses.
+	 * Otherwise, it closely resembles {@see wp_untrash_comment()}.
+	 *
+	 * @since [version]
+	 *
+	 * @param  WP_Comment|int $comment Comment (i.e., annotation) object or ID.
+	 *
+	 * @return bool                    True on success. False on failure.
+	 *
+	 * @see wp_untrash_comment()
+	 */
+	public static function untrash_comment( $comment ) {
+		if ( ! $comment ) {
+			return false;
+		}
+
+		$comment = get_comment( $comment );
+
+		if ( ! $comment ) {
+			return false;
+		}
+
+		/** This action is documented in wp-includes/comment.php */
+		do_action( 'untrash_comment', $comment->comment_ID, $comment );
+
+		$status = (string) get_comment_meta( $comment->comment_ID, '_wp_trash_meta_status', true );
+		$status = $status ? $status : '0';
+
+		if ( self::set_comment_status( $comment, $status ) ) {
+			delete_comment_meta( $comment->comment_ID, '_wp_trash_meta_time' );
+			delete_comment_meta( $comment->comment_ID, '_wp_trash_meta_status' );
+
+			/** This action is documented in wp-includes/comment.php */
+			do_action( 'untrashed_comment', $comment->comment_ID, $comment );
+
+			return true;
+		}
+
+		return false;
+	}
+
+	/**
+	 * Queries additional REST API collection parameters.
+	 *
+	 * @since [version]
+	 *
+	 * @param array           $query_vars {@see WP_Query} vars.
+	 * @param WP_REST_Request $request    REST API request.
+	 *
+	 * @return array                      Filtered query args.
+	 *
+	 * @see WP_Annotation_Utils::on_comments_clauses()
+	 * @see WP_REST_Comments_Controller::get_items()
+	 */
+	public static function on_rest_comment_query( $query_vars, $request ) {
```

Comment:
Cool, yes. I didn't realize class hints went all the way back to 5.0.

Reply:
I'm totally in favor of type hints anywhere they're supported, on just about everything. And I got excited when you mentioned that some work all the way back to 5.x. Woohoo! However, since this is not a practice that WordPress has adopted, and there's no mention of hints in the [coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/php/), I'm going to hold off for now and ask around. Maybe it's fine, I just haven't seen them used in core, have you?

Reply:
I asked about this on Slack and the impression I have is that it's probably OK to use type hints, but there needs to be some further discussion and a precedent set. Ideally, there would be something added to the WP code standards on this topic before we start using them.

Reply:
Don't worry about creating that Trac ticket @jaswrks I mentioned, this PR will suffice for said discussion 💯